File: admin/src/library/Socket.js
import Env from '@/env'
import Toast from '../library/ui/ui-toast'
import Storage from '../library/ui/storage'
import Notify from '../library/plus/notify'

/* Socket */
export default {

  self: null,
  store: null,

  /* 消息路由 */
  msgRouter(d){
    if(d.type=='group') this.msgGroup(d); //消息组
    else if(d.type=='msg') this.msg(d); //消息
    else if(d.type=='notify') this.msgNotify(d); //通知
  },

  /* 消息 */
  msg(d){
    // 阅读
    const voice = Storage.getItem('voice');
    Notify(d.data.title,d.data.content,voice?true:false);
  },

  /* 启动 */
  start(self){
    this.self = self;
    this.store = this.self.$store.state;
    // 重启Socket
    clearInterval(this.socketInterval);
    this.socketInterval = setInterval(()=>{
      if(this.store.isLogin && (!this.store.socket || this.store.socket.readyState!=1)) this.start(self);
    },Env.socket.time);
    // Token
    const token = Storage.getItem('token');
    if(!token) return false;
    // 开启
    if(Env.socket.start) this.socket(token);
  },

  /* 链接 */
  socket(token){
    this.store.socket = new WebSocket(Env.socket.server+'?token='+token);
    /* 链接 */
    this.store.socket.onopen = ()=>{
      console.log('Socket开启');
      // 心跳包
      clearInterval(this.heartbeat);
      this.heartbeat = setInterval(()=>{
        try{
          this.store.socket.send(JSON.stringify({type:''}));
        }catch(e){
          this._closeMsg();
        }
      },10000);
    }
    /* 关闭 */
    this.store.socket.onclose = ()=>{
      console.log('Socket关闭');
      this._closeMsg();
    }
    /* 接收 */
    this.store.socket.onmessage = (res)=>{
      const d = JSON.parse(res.data);
      // 是否成功
      if(d.code!=0) return Toast(d.msg);
      // 消息路由
      this.msgRouter(d);
    };
  },
  
  /* 关闭 */
  _closeMsg(){
    if(!this.store.socket) return;
    this.store.socket.close();
    this.store.socket = null;
  },

}

File: flutter/lib/library/Socket.dart
import 'package:flutter/material.dart';
import 'dart:convert' as convert;
import 'package:webmis/env.dart';
import 'package:webmis/library/inc/time-set.dart';
import 'package:webmis/library/ui/ui.dart';
import 'package:webmis/library/ui/storage.dart';
import 'package:webmis/library/ui/ui-toast.dart';
import 'package:webmis/library/plus/notify.dart';
import 'package:web_socket_channel/io.dart';
// import 'package:web_socket_channel/status.dart' as status;

/* Socket客户端 */
class Socket{

  static BuildContext _context;
  static var _channel;
  static var _socketInterval;
  static var _heartbeat;
  static var _store;

  /* 消息路由 */
  static Future<void> msgRouter(Map<String,dynamic> d) async {
    if(d['type']=='group') msgGroup(d); //消息组
    else if(d['type']=='msg') msg(d); //消息
    else if(d['type']=='notify') msgNotify(d); //通知
  }

  /* 消息组 */
  static Future<void> msgGroup(Map<String,dynamic> d) async {
    print(d);
  }

  /* 消息 */
  static Future<void> msg(Map<String,dynamic> d) async {
    // 阅读
    final voice = await Storage.getItem('voice');
    Notify(_context,d['data']['title'],d['data']['content'],isRead: voice==true??false);
    print(d);
  }

  /* 通知 */
  static Future<void> msgNotify(Map<String,dynamic> d) async {
    print(d);
  }

  /* 启动 */
  static Future<void> start(BuildContext context) async {
    _context = context;
    // 3秒刷新1次
    if(_socketInterval!=null) _socketInterval.cancel();
    _socketInterval = setInterval((res){
      _store = Ui.store(_context);
      Storage.setItem('token','EsA_l6u445xsZ53rPzTxzjlvVNQP4WG0m6mzfRxvbMOpd0ZNTZT9iTz20_AcG_QTb6LSJMvcTer4w8p9oq8j9qGTlebbmHBA9Gvb_CA=');
      Ui.store(_context).setIsLogin(true);
      // print('login: ${_store.isLogin} socket: ${_store.socket}');
      // 验证
      if(_store.isLogin && ( _store.socket==null || _channel==null )){
        Storage.getItem('token').then((token){
          socket(token);
        });
      }
    },3000);
  }

  /* 链接 */
  static Future socket(String token) async {
    _channel = IOWebSocketChannel.connect(Env.socket['server']+'?token='+token);
    print('Socket开启');
    Ui.store(_context).setSocket(_channel);
    // 心跳包
    if(_heartbeat!=null) _heartbeat.cancel();
    _heartbeat = setInterval((res){
      _channel.sink.add(convert.jsonEncode({'type':''}));
    },10000);
    // 接收
    _channel.stream.listen((msg) {
      Map<String,dynamic> d = convert.jsonDecode(msg);
      // 是否成功
      if(d['code']!=0) return Toast(_context,d['msg']);
      // 消息路由
      msgRouter(d);
    });
  }

}